libcaf_test/caf/test/approx.hpp

```diff
@@ -0,0 +1,54 @@
+// This file is part of CAF, the C++ Actor Framework. See the file LICENSE in
+// the main distribution directory for license terms and copyright or visit
+// https://github.com/actor-framework/actor-framework/blob/master/LICENSE.
+
+#pragma once
+
+#include <limits>
+#include <type_traits>
+
+namespace caf::test {
+
+/// Allows comparing a value of type `T` with a configurable precision.
+template <class T>
+class approx {
+public:
+  explicit approx(T value) : value_(value) {
+  }
+
+  approx epsilon(T eps) const {
+    auto copy = *this;
+    copy.epsilon_ = eps;
+    return copy;
+  }
+
+  friend bool operator==(T lhs, approx<T> rhs) {
+    if constexpr (std::is_floating_point_v<T>) {
+      return lhs >= rhs.min_value() && lhs <= rhs.max_value();
+    } else {
+      return lhs == rhs.value_;
+    }
+  }
+
+  friend bool operator>(T lhs, approx<T> rhs) {
+    if constexpr (std::is_floating_point_v<T>) {
+      return lhs > rhs.max_value();
+    } else {
+      return lhs > rhs.value_;
+    }
+  }
```

Comment:
Comparing against `approx` with `operator>` doesn't strike me as logically sound operation. Should be removed.

Reply:
I didn't check all places before where you've used `if constexpr` to check for floating points. I've assumed that we're dealing with arithmetic types, not with optionals and the like.

Ok, before we get lost in individual code pieces:

- `approx` should only work with types that are actual numbers, like integers and floating point values
- arrays, optionals, etc. should be dealt with at call side
- `approx` should only support `operator==`
- using `operator<` and `operator>` for floating points is fine, we only want to prohibit `operator==` for floating point numbers. The right place to put the `static_assert` might be just `check_eq`.

Reply:
Thanks, this clarifies things up a lot. I am updating the PR to ensure that these are enforced properly.